### PR TITLE
add logic to handle deletion of VRG CR

### PR DESCRIPTION
**Pull request by raghavendrabhat**
* Add finalizer to VRG CR.

Signed-off-by: Raghavendra M <raghavendra@redhat.com>

### CODE REVIEW

File: controllers/volumereplicationgroup_controller.go
Comment (BenamarMk) [inline controllers/volumereplicationgroup_controller.go:280]: I suggest moving this line outside RemoveFinalizer().  In other words, let RemoveFinalizer only removes the finalizer from the K8s object.  A possible change would be made in line 212 and would be something something like:
```
         if containsString(volRepGroup.ObjectMeta.Finalizers, finalizer) {
		if err := v.deleteRelatedItems(ctx, volRepGroup); err != nil {
		        ........
		}
        }

        if err := v.RemoveFinalizer(ctx, volRepGroup, vrgFinalizer); err != nil {
			return ctrl.Result{}, fmt.Errorf("failed to remove the finalizer from VRG CR %s (%w)", req.NamespacedName, err)
		}
```

File: controllers/volumereplicationgroup_controller.go
Comment (BenamarMk) [inline controllers/volumereplicationgroup_controller.go:322]: Maybe return here as well rather than falling through

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:280]: > I suggest moving this line outside RemoveFinalizer(). In other words, let RemoveFinalizer only removes the finalizer from the K8s object. A possible change would be made in line 212 and would be something something like:
> 
> ```
>          if containsString(volRepGroup.ObjectMeta.Finalizers, finalizer) {
> 		if err := v.deleteRelatedItems(ctx, volRepGroup); err != nil {
> 		        ........
> 		}
>         }
> 

So, line 212 is from Reconcile function and its length would exceed 60 lines failing the linter check. Alternatively what I can do is this:

- Let RemoveFinalizer only remove the finalizer.
- Add another function (say CheckAndRemoveFinalier) which would do the checking of whether the finalizer is
  there or not and if present, then call RemoveFinalizer.
- From reconcile (i.e. line 212) call "CheckAndRemoveFinalizer"

Does it look ok?


>         if err := v.RemoveFinalizer(ctx, volRepGroup, vrgFinalizer); err != nil {
> 			return ctrl.Result{}, fmt.Errorf("failed to remove the finalizer from VRG CR %s (%w)", req.NamespacedName, err)
> 		}
> ```

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:322]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (BenamarMk) [reply controllers/volumereplicationgroup_controller.go:280]: That works

File: controllers/volumereplicationgroup_controller.go
Comment (vdeenadh) [inline controllers/volumereplicationgroup_controller.go:55]: For improved readability in places where it is being used, to clarify that this is merely the name of the finalizer, can we add a "Name" suffix, vrgFinalizerName?

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [inline controllers/volumereplicationgroup_controller.go:483]: (nit) Add as a comment to this function, instead of in the body?

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [inline controllers/volumereplicationgroup_controller.go:481]: `volRepGroup` unused in this function. (linter should catch these :( )

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [inline controllers/volumereplicationgroup_controller.go:493]: (future) as the comment above states. We could possibly create VR and backup the PV, per PVC instead of waiting for all VRs to be created before backing the PVs up.

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [reply controllers/volumereplicationgroup_controller.go:306]: Yes, VR marked as `Secondary` and later deleted should be a pre-requisite to consider VRG as deleted and work completed. I think @raghavendrabhat is planning on a subsequent PR to address this?

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [reply controllers/volumereplicationgroup_controller.go:293]: Yes, we should only operate on VR resources that we created and hence manage. Possibly add in a subsequent PR?

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [reply controllers/volumereplicationgroup_controller.go:481]: Ah! the [linter](https://github.com/mvdan/unparam#unparam) does catch it, but allows it for exported functions (which is the case here as of this version).

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:55]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:483]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:478]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:483]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:478]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [reply controllers/volumereplicationgroup_controller.go:288]: If we never got a chance to add a finalizer we will not see one on delete, and so no work to do to cleanup from our end. There maybe other finalizers that prevent deletion and hence we get the delete event. So this is fine, no else condition is required.

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:207]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:209]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:239]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:267]: ACK. Removed the background and simply using context

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:275]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:280]: ACK

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:293]: ACK. Additional checks to handle only those VR resources created by VRG will be done in a subsequent PR.

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:306]: Yes. This will be done as a separate PR.

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:310]: ACK. Renaming the functions.

instead of using VR or VRCR "VolumeReplicationResource" is used where possible.

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:337]: Yes. Additional logic to delete only those VR resources created by VRG would be handled in a subsequenet PR.

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:460]: Yes. Handling only those VR resources created by VRG will be part of a subsequent PR.

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:493]: Yes. This is deferred to a separate PR.

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [inline controllers/volumereplicationgroup_controller.go:55]: Should we name this `volumereplicationgroups.ramendr.openshift.io/vrg-protection`? Going along the lines of <API-group>/<name>

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [inline controllers/volumereplicationgroup_controller.go:501]: Can reduce one level of nesting if we process the no-errors first. if handling the other changes can consider this as well.
```suggestion
		if err != nil {
```
```suggestion
		if err == nil {
		    continue
		    }
```

**Review by ShyamsundarR — COMMENTED** [review]
Mostly LGTM!

Check finalizer name if it is better and relates to our API endpoint name as well.

File: controllers/volumereplicationgroup_controller.go
Comment (BenamarMk) [inline controllers/volumereplicationgroup_controller.go:209]: So we are queuing without any logging.  Either log the reason why you are requeuing or `return ctrl.Result{}, err`

File: controllers/volumereplicationgroup_controller.go
Comment (BenamarMk) [inline controllers/volumereplicationgroup_controller.go:218]: Same here; please log it or `return ctrl.Result{}, err`

File: controllers/volumereplicationgroup_controller.go
Comment (BenamarMk) [inline controllers/volumereplicationgroup_controller.go:250]: (nit) Name result parameters are confusing and there is no reason to not declare a variable inside the function.  This looks weird.

File: controllers/volumereplicationgroup_controller.go
Comment (BenamarMk) [inline controllers/volumereplicationgroup_controller.go:407]: (nit) I prefer to use `return !requeue`.  It reads better when the previously you had `return requeue` and then the last one is `return false`.  So you may need to declar requeue outside the `if`

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:250]: ACK. Will change it.

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:407]: From your suggestion it seems you are telling me to declare requeue (initialized to true?) in the beginning and then at the
end "return !requeue".

That might create some issues. Assume requeue is initialized to true at the time of declaration. Then both of the if statements are false. So ideally we should be returning false from the function. But doing "return !requeue" leads to returning true.

What I can do is, initialize requeue to false at the time of declaration and at the end do just "return requeue". That way, if any of the if conditions are true, then immediately, requeue is set to true and that is returned then and there itself. By the time we reach end of the function, both the if conditions are false, hence requeue would still be false and safely we can return it.

NOTE: I am not sure about scope of the values returned in if blocks for the variables that are declared outside the if blocks  
            in Go. 

Let me know your thoughts.

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:218]: ACK.

File: controllers/volumereplicationgroup_controller.go
Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:209]: ACK. Will log the reason.

File: controllers/volumereplicationgroup_controller.go
Comment (BenamarMk) [reply controllers/volumereplicationgroup_controller.go:407]: The reason I suggested the change is because these requeues are coming from everywhere and it becomes easily messy, especially when the linter forces you to split a function.  What I suggested was very simple.  Here it is:
```
func (v *VolumeReplicationGroupReconciler) handlePersistentVolumeClaims(
	ctx context.Context,
	volRepGroup *ramendrv1alpha1.VolumeReplicationGroup,
	pvcList *corev1.PersistentVolumeClaimList) bool {

        const requeue = true

	if requeueAction := v.createVolumeReplicationResources(ctx, volRepGroup, pvcList); requeueAction {
		log.Error("failed to get or create VolumeReplication CRs for pvc list")

		return requeue
	}

	if requeueAction := v.handlePersistentVolumes(ctx, pvcList); requeueAction {
		log.Error("failed to handle Persistent Volumes for pvc list")

		return requeue
	}

	return !requeue
}
```
That's all.  I named the tmp variable requeueAction just to show what I mean.  You can name that anything meaningful and even `rq`
But again, it was a small suggestion only and it will not hold up the PR.

Comment (raghavendrabhat) [reply controllers/volumereplicationgroup_controller.go:407]: ACK.

File: controllers/volumereplicationgroup_controller.go
Comment (ShyamsundarR) [reply controllers/volumereplicationgroup_controller.go:55]: I see that my comment was cut off now... this is fine, I was looking at the name as `volumereplicationgroups.ramendr.openshift.io/vrg-protection` we can change it subsequently.

**Review by ShyamsundarR — APPROVED** [review]
Looks good to proceed. @BenamarMk or @vdeenadh if fine we can merge this and proceed?

**Comment by raghavendrabhat** [comment]
> I see that my comment was cut off now... this is fine, I was looking at the name as volumereplicationgroups.ramendr.openshift.io/vrg-protection we can change it subsequently.

Ahh. Sorry I missed this. I mean I missed to respond. I agree with you. I was having confusion with the name and thought of doing it as a change in a subsequent PR.

**Review by BenamarMk — APPROVED** [review]
@raghavendrabhat is this ready to merge?

**Comment by ShyamsundarR** [comment]
> @raghavendrabhat is this ready to merge?

@BenamarMk ack, was waiting for another approval, merging now.

**Pull request merged**
